Packages/com.unity.inputsystem/InputSystem/Plugins/Visualizers/VisualizationHelpers.cs

```diff
@@ -0,0 +1,623 @@
+                samplesText = new ArrayHelpers.RingBuffer<GUIContent>(numSamples);
+            }
+
+            public override void AddSample(object value, double time)
+            {
+                var v = default(TValue);
+
+                if (value != null)
+                {
+                    if (!(value is TValue val))
+                        throw new ArgumentException(
+                            $"Expecting value of type '{typeof(TValue).Name}' but value of type '{value?.GetType().Name}' instead",
+                            nameof(value));
+                    v = val;
+                }
+
+                samples.Append(v);
+                samplesText.Append(new GUIContent(v.ToString()));
+            }
+        }
+
+        // Visualizes integer and real type primitives.
+        public class ScalarVisualizer<TValue> : ValueVisualizer<TValue>
+            where TValue : struct
+        {
+            public TValue limitMin;
+            public TValue limitMax;
+            public TValue min;
+            public TValue max;
+
+            public ScalarVisualizer(int numSamples = 10)
+                : base(numSamples)
+            {
+            }
+
+            public override void OnDraw(Rect rect)
+            {
+                // For now, only draw the current value.
+                DrawRectangle(rect, new Color(1, 1, 1, 0.1f));
+                if (samples.count == 0)
+                    return;
+                var sample = samples[samples.count - 1];
+                if (Compare(sample, default) == 0)
+                    return;
+                if (Compare(limitMin, default) != 0)
+                {
+                    // Two-way visualization with positive and negative side.
+                    throw new NotImplementedException();
+                }
+                else
+                {
+                    // One-way visualization with only positive side.
+                    var ratio = Divide(sample, limitMax);
+                    var fillRect = rect;
+                    fillRect.width = rect.width * ratio;
+                    DrawRectangle(fillRect, new Color(0, 1, 0, 0.75f));
+
+                    var valuePos = new Vector2(fillRect.xMax, fillRect.y + fillRect.height / 2);
+                    DrawText(samplesText[samples.count - 1], valuePos, ValueTextStyle);
+                }
+            }
+
+            public override void AddSample(object value, double time)
+            {
+                base.AddSample(value, time);
+
+                if (value != null)
+                {
+                    var val = (TValue)value;
+                    if (Compare(min, val) > 0)
+                        min = val;
+                    if (Compare(max, val) < 0)
+                        max = val;
+                }
+            }
+
+            private static unsafe int Compare(TValue left, TValue right)
+            {
+                var leftPtr = UnsafeUtility.AddressOf(ref left);
```

Comment:
why do we need the unsafe logic here? Couldn't we just do something like
```if (typeof(TValue) == typeof(int) || typeof(TValue) == typeof(float))  
return left.CompareTo(right);```
?

Reply:
I wish. C# won't allow that. `as` can only be used on reference and nullable types. With the current level of type parameter constraints in C#, I found no other way to do this kind of stuff more easily and without garbage.

Reply:
meh, ok.